File: internal/storage/storage.go
// Package storage is where haul keeps its config, logins and download archive: $HAUL_HOME, else
// $XDG_CONFIG_HOME/haul, else ~/.config/haul.
package storage

import (
	"os"
	"path/filepath"
	"strings"
)

const (
	CookieFile   = "cookie.txt"
	TVTokenFile  = "tv-token.txt"
	AppTokenFile = "app-token.txt"
	ConfigFile   = "config.json"
	ArchiveFile  = "archives.txt"
)

// Home is the directory.
func Home() string {
	if h := os.Getenv("HAUL_HOME"); h != "" {
		return h
	}
	if x := os.Getenv("XDG_CONFIG_HOME"); x != "" {
		return filepath.Join(x, "haul")
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return filepath.Join(".config", "haul")
	}
	return filepath.Join(home, ".config", "haul")
}

// Path is a file in it.
func Path(name string) string { return filepath.Join(Home(), name) }

// Read is a file's trimmed text; "" when missing or empty.
func Read(name string) string {
	b, err := os.ReadFile(Path(name))
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(b))
}

// Write replaces a file, readable by its owner only: it may hold a login.
func Write(name, text string) error {
	if err := os.MkdirAll(Home(), 0o755); err != nil {
		return err
	}
	return os.WriteFile(Path(name), []byte(text), 0o600)
}

// Append adds text at the end of a file.
func Append(name, text string) error {
	if err := os.MkdirAll(Home(), 0o755); err != nil {
		return err
	}
	f, err := os.OpenFile(Path(name), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = f.WriteString(text)
	return err
}

File: Sources/HaulCore/Core/Storage.swift
import Foundation

/// Where credentials, config and the download archive live: `$HAUL_HOME`, else `$XDG_CONFIG_HOME/haul`, else `~/.config/haul`.
public enum Storage {
    public static let cookieFile = "cookie.txt"
    public static let tvTokenFile = "tv-token.txt"
    public static let appTokenFile = "app-token.txt"
    public static let configFile = "config.json"
    public static let archiveFile = "archives.txt"

    public static var home: URL {
        let env = ProcessInfo.processInfo.environment
        if let h = env["HAUL_HOME"], !h.isEmpty { return URL(fileURLWithPath: h) }
        if let x = env["XDG_CONFIG_HOME"], !x.isEmpty { return URL(fileURLWithPath: x).appendingPathComponent("haul") }
        return FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".config/haul")
    }

    public static func url(_ name: String) -> URL { home.appendingPathComponent(name) }

    public static func read(_ name: String) -> String? {
        guard let s = try? String(contentsOf: url(name), encoding: .utf8) else { return nil }
        let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }

    public static func write(_ name: String, _ text: String) throws {
        try FileManager.default.createDirectory(at: home, withIntermediateDirectories: true)
        try text.write(to: url(name), atomically: true, encoding: .utf8)
    }

    public static func append(_ name: String, _ text: String) throws {
        try FileManager.default.createDirectory(at: home, withIntermediateDirectories: true)
        let u = url(name)
        if !FileManager.default.fileExists(atPath: u.path) { try "".write(to: u, atomically: true, encoding: .utf8) }
        let h = try FileHandle(forWritingTo: u)
        defer { try? h.close() }
        try h.seekToEnd()
        try h.write(contentsOf: Data(text.utf8))
    }
}
